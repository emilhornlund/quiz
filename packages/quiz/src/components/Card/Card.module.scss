@use '../../styles/colors';
@use '../../styles/helpers';
@use '../../styles/variables';

.card {
  display: flex;
  flex-direction: column;
  border-style: solid;
  position: relative;

  &.kindPrimary {
    border-color: colors.$white;

    .content {
      background-color: colors.$white;
      color: colors.$gray-4;
    }
  }

  &.kindCallToAction {
    border-color: colors.$yellow-2;

    .content {
      background-color: colors.$yellow-2;
      color: colors.$gray-4;
    }
  }

  &.kindSuccess {
    border-color: colors.$green-2;

    .content {
      background-color: colors.$green-2;
      color: colors.$white;
    }
  }

  &.sizeSmall {
    @include helpers.mobile {
      width: 100%;
    }
    @include helpers.tablet {
      width: variables.$tablet-width * 0.4;
    }
    @include helpers.desktop {
      width: variables.$desktop-width * 0.4;
    }
  }

  &.sizeMedium {
    @include helpers.mobile {
      width: 100%;
    }
    @include helpers.tablet {
      width: variables.$tablet-width * 0.6;
    }
    @include helpers.desktop {
      width: variables.$desktop-width * 0.6;
    }
  }

  &.sizeLarge {
    @include helpers.mobile {
      width: 100%;
    }
    @include helpers.tablet {
      width: variables.$tablet-width * 0.8;
    }
    @include helpers.desktop {
      width: variables.$desktop-width * 0.8;
    }
  }

  &.sizeFull {
    width: 100%;
  }

  &.center {
    .content {
      align-items: center;
      justify-content: center;
      text-align: center;
    }
  }

  .content {
    display: flex;
    flex-direction: column;

    a {
      font-style: italic;
      font-weight: bold;
      color: colors.$gray-4;
      cursor: pointer;
    }
  }

  @include helpers.mobile {
    padding: helpers.pxToRem(2px);
    border-radius: variables.$mobile-control-normal-border-radius;
    border-width: helpers.pxToRem(2px);

    .content {
      padding: variables.$mobile-narrow-layout-gutter;
      border-radius: calc(#{variables.$mobile-control-normal-border-radius} / 2);
      line-height: helpers.pxToRem(26px);

      @include helpers.fontSize(16px);
    }
  }

  @include helpers.tablet {
    padding: helpers.pxToRem(2px);
    border-radius: variables.$tablet-control-normal-border-radius;
    border-width: helpers.pxToRem(2px);

    .content {
      padding: variables.$tablet-narrow-layout-gutter;
      border-radius: calc(#{variables.$tablet-control-normal-border-radius} / 2);
      line-height: helpers.pxToRem(28px);

      @include helpers.fontSize(18px);
    }
  }

  @include helpers.desktop {
    padding: helpers.pxToRem(3px);
    border-radius: variables.$desktop-control-normal-border-radius;
    border-width: helpers.pxToRem(3px);

    .content {
      padding: variables.$desktop-narrow-layout-gutter;
      border-radius: calc(#{variables.$desktop-control-normal-border-radius} / 2);
      line-height: helpers.pxToRem(30px);

      @include helpers.fontSize(20px);
    }
  }

  button {
    appearance: none;
    outline: none;
    border: none;
    background-color: colors.$white;
    color: colors.$gray-4;
    cursor: pointer;
    position: absolute;

    @include helpers.mobile {
      top: calc(#{helpers.pxToRem(11px) * -0.85});
      right: calc(#{helpers.pxToRem(11px) * -0.85});
      width: helpers.pxToRem(22px);
      height: helpers.pxToRem(22px);
      border-radius: helpers.pxToRem(22px);

      @include helpers.fontSize(14px);
    }

    @include helpers.tablet {
      top: calc(#{helpers.pxToRem(13px) * -0.85});
      right: calc(#{helpers.pxToRem(13px) * -0.85});
      width: helpers.pxToRem(26px);
      height: helpers.pxToRem(26px);
      border-radius: helpers.pxToRem(13px);

      @include helpers.fontSize(16px);
    }

    @include helpers.desktop {
      top: calc(#{helpers.pxToRem(15px) * -0.85});
      right: calc(#{helpers.pxToRem(15px) * -0.85});
      width: helpers.pxToRem(30px);
      height: helpers.pxToRem(30px);
      border-radius: helpers.pxToRem(15px);

      @include helpers.fontSize(18px);
    }

    &:hover {
      color: colors.$gray-2;
    }
  }
}
