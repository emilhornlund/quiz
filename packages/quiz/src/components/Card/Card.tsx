import { faXmark } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import React, { FC, MouseEvent } from 'react'

import { classNames } from '../../utils/helpers.ts'

import styles from './Card.module.scss'

export type CardKind = 'primary' | 'call-to-action' | 'success'

export type CardSize = 'small' | 'medium' | 'large' | 'full'

const CardKindClassName: {
  [key in CardKind]: string
} = {
  primary: styles.kindPrimary,
  'call-to-action': styles.kindCallToAction,
  success: styles.kindSuccess,
}

const CardSizeClassName: { [key in CardSize]: string } = {
  small: styles.sizeSmall,
  medium: styles.sizeMedium,
  large: styles.sizeLarge,
  full: styles.sizeFull,
}

export interface CardProps {
  kind?: CardKind
  size?: CardSize
  center?: boolean
  onDismiss?: () => void
  children?: React.ReactNode
}

const Card: FC<CardProps> = ({
  kind = 'primary',
  size = 'full',
  center,
  onDismiss,
  children,
}) => {
  const handleDismissClick = (event: MouseEvent): void => {
    event?.preventDefault()
    onDismiss?.()
  }

  return (
    <div
      className={classNames(
        styles.card,
        CardKindClassName[kind],
        CardSizeClassName[size],
        center ? styles.center : undefined,
      )}>
      {onDismiss && (
        <button type="button" onClick={handleDismissClick}>
          <FontAwesomeIcon icon={faXmark} />
        </button>
      )}
      <div className={styles.content}>{children}</div>
    </div>
  )
}

export default Card
