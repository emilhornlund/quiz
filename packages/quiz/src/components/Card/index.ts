export type { CardKind, CardProps, CardSize } from './Card'
export { default } from './Card'
