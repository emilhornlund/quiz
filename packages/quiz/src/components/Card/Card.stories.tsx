import type { Meta, StoryObj } from '@storybook/react'
import React from 'react'

import Card from './Card'

const meta = {
  title: 'Surfaces/Card',
  component: Card,
  tags: ['autodocs'],
} satisfies Meta<typeof Card>

export default meta
type Story = StoryObj<typeof meta>

const EXAMPLE_TEXT =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. '

export const Primary = {
  args: {
    kind: 'primary',
    center: false,
    children: <>{EXAMPLE_TEXT}</>,
  },
} satisfies Story

export const CallToAction = {
  args: {
    kind: 'call-to-action',
    center: false,
    children: <>{EXAMPLE_TEXT}</>,
  },
} satisfies Story

export const Success = {
  args: {
    kind: 'success',
    center: false,
    children: <>{EXAMPLE_TEXT}</>,
  },
} satisfies Story
