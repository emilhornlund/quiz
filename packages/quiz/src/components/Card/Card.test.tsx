import { render } from '@testing-library/react'
import React from 'react'
import { describe, expect, it } from 'vitest'

import Card from './Card'

describe('Card', () => {
  it('should render a Card with default props', async () => {
    const { container } = render(<Card />)

    expect(container).toMatchSnapshot()
  })

  it('should render a Card with prop kind primary', async () => {
    const { container } = render(<Card kind="primary" />)

    expect(container).toMatchSnapshot()
  })

  it('should render a Card with prop kind call-to-action', async () => {
    const { container } = render(<Card kind="call-to-action" />)

    expect(container).toMatchSnapshot()
  })

  it('should render a Card with prop kind success', async () => {
    const { container } = render(<Card kind="success" />)

    expect(container).toMatchSnapshot()
  })

  it('should render a Card with prop size small', async () => {
    const { container } = render(<Card size="small" />)

    expect(container).toMatchSnapshot()
  })

  it('should render a Card with prop size medium', async () => {
    const { container } = render(<Card size="medium" />)

    expect(container).toMatchSnapshot()
  })

  it('should render a Card with prop size large', async () => {
    const { container } = render(<Card size="large" />)

    expect(container).toMatchSnapshot()
  })

  it('should render a Card with prop size full', async () => {
    const { container } = render(<Card size="full" />)

    expect(container).toMatchSnapshot()
  })

  it('should render a Card with prop center true', async () => {
    const { container } = render(<Card center />)

    expect(container).toMatchSnapshot()
  })

  it('should render a Card with prop center false', async () => {
    const { container } = render(<Card center={false} />)

    expect(container).toMatchSnapshot()
  })

  it('should render a Card with children', async () => {
    const { container } = render(<Card>some text</Card>)

    expect(container).toMatchSnapshot()
  })
})
