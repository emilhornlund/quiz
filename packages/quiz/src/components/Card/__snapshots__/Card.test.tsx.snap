// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card > should render a Card with children 1`] = `
<div>
  <div
    class="card kindPrimary sizeFull"
  >
    <div
      class="content"
    >
      some text
    </div>
  </div>
</div>
`;

exports[`Card > should render a Card with default props 1`] = `
<div>
  <div
    class="card kindPrimary sizeFull"
  >
    <div
      class="content"
    />
  </div>
</div>
`;

exports[`Card > should render a Card with prop center false 1`] = `
<div>
  <div
    class="card kindPrimary sizeFull"
  >
    <div
      class="content"
    />
  </div>
</div>
`;

exports[`Card > should render a Card with prop center true 1`] = `
<div>
  <div
    class="card kindPrimary sizeFull center"
  >
    <div
      class="content"
    />
  </div>
</div>
`;

exports[`Card > should render a Card with prop kind call-to-action 1`] = `
<div>
  <div
    class="card kindCallToAction sizeFull"
  >
    <div
      class="content"
    />
  </div>
</div>
`;

exports[`Card > should render a Card with prop kind primary 1`] = `
<div>
  <div
    class="card kindPrimary sizeFull"
  >
    <div
      class="content"
    />
  </div>
</div>
`;

exports[`Card > should render a Card with prop kind success 1`] = `
<div>
  <div
    class="card kindSuccess sizeFull"
  >
    <div
      class="content"
    />
  </div>
</div>
`;

exports[`Card > should render a Card with prop size full 1`] = `
<div>
  <div
    class="card kindPrimary sizeFull"
  >
    <div
      class="content"
    />
  </div>
</div>
`;

exports[`Card > should render a Card with prop size large 1`] = `
<div>
  <div
    class="card kindPrimary sizeLarge"
  >
    <div
      class="content"
    />
  </div>
</div>
`;

exports[`Card > should render a Card with prop size medium 1`] = `
<div>
  <div
    class="card kindPrimary sizeMedium"
  >
    <div
      class="content"
    />
  </div>
</div>
`;

exports[`Card > should render a Card with prop size small 1`] = `
<div>
  <div
    class="card kindPrimary sizeSmall"
  >
    <div
      class="content"
    />
  </div>
</div>
`;
